bson/src/main/org/bson/internal/CodecCache.java

```diff
@@ -19,36 +19,75 @@
 import org.bson.codecs.Codec;
 import org.bson.codecs.configuration.CodecConfigurationException;
 
+import java.lang.reflect.Type;
+import java.util.List;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
 
 import static java.lang.String.format;
 
 final class CodecCache {
-    private final ConcurrentMap<Class<?>, Optional<Codec<?>>> codecCache = new ConcurrentHashMap<>();
 
-    public boolean containsKey(final Class<?> clazz) {
-        return codecCache.containsKey(clazz);
+    static final class CodecCacheKey {
+        private final Class<?> clazz;
+        private final List<Type> types;
+
+        CodecCacheKey(final Class<?> clazz, final List<Type> types) {
+            this.clazz = clazz;
+            this.types = types;
+        }
+
+        @Override
+        public boolean equals(final Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            CodecCacheKey that = (CodecCacheKey) o;
+            return clazz.equals(that.clazz) && Objects.equals(types, that.types);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(clazz, types);
+        }
+
+        @Override
+        public String toString() {
+            return "CodecCacheKey{"
+                    + "clazz=" + clazz
+                    + ", types=" + types
+                    + '}';
+        }
+    }
+
+    private final ConcurrentMap<CodecCacheKey, Optional<Codec<?>>> codecCache = new ConcurrentHashMap<>();
+
+    public boolean containsKey(final CodecCacheKey codecCacheKey) {
+        return codecCache.containsKey(codecCacheKey);
     }
 
-    public void put(final Class<?> clazz, final Codec<?> codec){
-        codecCache.put(clazz, Optional.ofNullable(codec));
+    public void put(final CodecCacheKey codecCacheKey, final Codec<?> codec){
+        codecCache.put(codecCacheKey, Optional.ofNullable(codec));
     }
 
     @SuppressWarnings("unchecked")
-    public synchronized <T> Codec<T> putIfMissing(final Class<T> clazz, final Codec<T> codec) {
-        Optional<Codec<?>> cachedCodec = codecCache.computeIfAbsent(clazz, clz -> Optional.of(codec));
+    public synchronized <T> Codec<T> putIfMissing(final CodecCacheKey codecCacheKey, final Codec<T> codec) {
+        Optional<Codec<?>> cachedCodec = codecCache.computeIfAbsent(codecCacheKey, clz -> Optional.of(codec));
         if (cachedCodec.isPresent()) {
             return (Codec<T>) cachedCodec.get();
         }
-        codecCache.put(clazz, Optional.of(codec));
+        codecCache.put(codecCacheKey, Optional.of(codec));
```

Comment:
Something appears to be wrong with how this method was implemented before the PR, and how it is implemented now. It likely can be implemented as follows (note that `synchronized` is no longer there):

```java
@SuppressWarnings("unchecked")
public <T> Codec<T> putIfMissing(final CodecCacheKey codecCacheKey, final Codec<T> codec) {
    return codecCache.compute(codecCacheKey, (k, v) -> {
        if (v != null && v.isPresent()) {
            return v;
        } else {
            return Optional.of(codec);
        }}).map(c -> (Codec<T>)c).orElseThrow(() -> new AssertionError("unreachable"));
}
```

Tangentially related to the above improvement, not to this PR, methods `OverridableUuidRepresentationCodecRegistry.get`, `ProvidersCodecRegistry.get` do two hash map lookups instead of one on the happy path, because they both use the anti-pattern "check if a binding is there and then find the binding again, this time return the value".

Reply:
Wrong as in both readability and performance could be better. I was initially baffled by seeing a put in a `ConcurrentMap` that is guarded with `synchronized` and is implemented as two puts for the same key.

Reply:
In that case, let's fix this in a separate PR.

Reply:
I forgot to link the new ticket here: https://jira.mongodb.org/browse/JAVA-4751. Will do when this PR is merged to avoid conflicts.